cloudinit/sources/DataSourceAzure.py

```diff
@@ -660,10 +691,249 @@ def setup(self, is_new_instance):
+
+            LOG.info("Unbinding and binding the interface %s", ifname)
+            devicename = net.read_sys_net(ifname,
+                                          'device/device_id').strip('{}')
+            util.write_file('/sys/bus/vmbus/drivers/hv_netvsc/unbind',
+                            devicename)
+            util.write_file('/sys/bus/vmbus/drivers/hv_netvsc/bind',
+                            devicename)
+
+            attempts = attempts + 1
+            if self.distro.networking.try_set_link_up(ifname):
+                msg = "The link %s is up after %s attempts" % (ifname,
+                                                               attempts)
+                report_diagnostic_event(msg, logger_func=LOG.info)
+                return
+
+            LOG.info("Link is not up despite attempts to bring it up. "
+                     "Sleeping for 1s and trying again.")
+            sleep(1)
+
+    @azure_ds_telemetry_reporter
+    def _create_report_ready_marker(self):
+        path = REPORTED_READY_MARKER_FILE
+        LOG.info(
+            "Creating a marker file to report ready: %s", path)
+        util.write_file(path, "{pid}: {time}\n".format(
+            pid=os.getpid(), time=time()))
+        report_diagnostic_event(
+            'Successfully created reported ready marker file '
+            'while in the preprovisioning pool.',
+            logger_func=LOG.debug)
+
+    @azure_ds_telemetry_reporter
+    def _report_ready_if_needed(self):
+        """Report ready to the platform if the marker file is not present,
+        and create the marker file.
+        """
+        have_not_reported_ready = bool(
+            not os.path.isfile(REPORTED_READY_MARKER_FILE))
+
+        if have_not_reported_ready:
+            LOG.info("Reporting ready to platform")
+            try:
+                with EphemeralDHCPv4WithReporting(azure_ds_reporter) as lease:
+                    self._report_ready(lease=lease)
+            except Exception as e:
+                msg = "Exception reporting ready during preprovisioning: %s" %\
+                    e
+                report_diagnostic_event(msg, logger_func=LOG.error)
+                raise
+            self._create_report_ready_marker()
+
+    @azure_ds_telemetry_reporter
+    def _check_if_nic_is_primary(self, ifname):
+        """Check if a given interface is the primary nic or not. If it is the
+        primary nic, then we also get the expected total nic count from IMDS.
+        """
+        is_primary = False
+        expected_nic_count = -1
+        imds_md = None
+
+        # We don't need to do dhcp on all NICs. Cloud-init
+        # only does dhcp on the primary/fallback NIC. Since
+        # we do not yet know which is the primary NIC, we
+        # do dhcp on each NIC until we find it.
+        with events.ReportEventStack(
+                name="obtain-dhcp-lease",
+                description="obtain dhcp lease",
+                parent=azure_ds_reporter):
```

Comment:
Agree with @OddBloke. The `ifname` should be logged in the `description` of the `EventStack`. Ideally, the `description` field should also say `obtain dhcp lease when attempting to determine primary NIC during PPS VM restore workflow`.

Note: Internal telemetry systems rely on the EventStack `name` field being `obtain-dhcp-lease` to detect in case there is a DHCP failure.